konnectivity-client/pkg/client/client.go

```diff
@@ -185,14 +189,41 @@ func CreateSingleUseGrpcTunnelWithContext(createCtx, tunnelCtx context.Context,
 }
 
 func newUnstartedTunnel(stream client.ProxyService_ProxyClient, c clientConn) *grpcTunnel {
-	return &grpcTunnel{
+	t := grpcTunnel{
 		stream:             stream,
 		clientConn:         c,
 		pendingDial:        pendingDialManager{pendingDials: make(map[int64]pendingDial)},
 		conns:              connectionManager{conns: make(map[int64]*conn)},
 		readTimeoutSeconds: 10,
 		done:               make(chan struct{}),
+		prevStatus:         metrics.TunnelStatusCreated,
 	}
+	t.initMetric(metrics.TunnelStatusCreated)
+	return &t
+}
+
+// initMetric must be called exactly once to initialize tunnels metric.
+func (t *grpcTunnel) initMetric(status metrics.TunnelStatus) {
+	t.statusLock.Lock()
+	defer t.statusLock.Unlock()
+	metrics.Metrics.GetTunnelsMetric().WithLabelValues(string(status)).Inc()
+	t.prevStatus = status
+}
+
+func (t *grpcTunnel) updateMetric(status metrics.TunnelStatus) {
+	t.statusLock.Lock()
+	defer t.statusLock.Unlock()
+	m := metrics.Metrics.GetTunnelsMetric()
+	m.WithLabelValues(string(t.prevStatus)).Dec()
```

Comment:
Dislike having this much under lock, especially when we are acquiring yet more locks under the statusLock. 

Maybe something more like 
func (t *grpcTunnel) setMetric(status metrics.TunnelStatus) metrics.TunnelStatus {
	t.statusLock.Lock()
	defer t.statusLock.Unlock()
	prev := t.prevStatus
	t.prevStatus = status
	return prev
}

func (t *grpcTunnel) updateMetric(status metrics.TunnelStatus) {
	prev := t.setMetric(status)
	m := metrics.Metrics.GetTunnelsMetric()
	m.WithLabelValues(string(prev)).Dec()
	m.WithLabelValues(string(status)).Inc()
}

Reply:
Used atomic.Value.